File: Leetcode 0039 Combination Sum [medium] [array]/Combination Sum [dfs].java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        if (candidates == null) {
            return result;
        }
        
        List<Integer> combination = new ArrayList<>();
        Arrays.sort(candidates);
        dfs(candidates, target, result, combination, 0);
        return result;
    }
    
    private void dfs(int[] candidates, int target, List<List<Integer>> result, List<Integer> combination, int index) {
        if (target < 0) {
            return;
        }
        
        if (target == 0) {
            result.add(new ArrayList<Integer>(combination));
            return;
        }
        
        for (int i = index; i < candidates.length; i++) {
            combination.add(candidates[i]);
            dfs(candidates, target - candidates[i], result, combination, i);
            combination.remove(combination.size() - 1);
        }
    }
}

File: Leetcode 0039 Combination Sum [medium] [array]/Combination Sum [dfs].js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    let result = [];
    if (candidates === null || candidates.length === 0) {
        return result;
    }
    
    let combination = [];
    candidates.sort((a, b) => a - b);
    dfs(candidates, target, result, combination, 0);
    return result;
};

let dfs = (candidates, target, result, combination, index) => {
    if (target < 0) {
        return;
    }
    
    if (target === 0) {
        result.push(combination.slice(0));
        return;
    }
    
    for (let i = index; i < candidates.length; i++) {
        combination.push(candidates[i]);
        dfs(candidates, target - candidates[i], result, combination, i);
        combination.pop();
    }
}